### PR TITLE
raw secure storage accesses

### PR DESCRIPTION
**Pull request by etienne-lms**
secure_storage example provides basics for creating, reading and deleting
a object in a TA secure storage.

A TA command allows to create an object in the TA secure storage.
A TA command allows to read an object in the TA secure storage.
A TA command allows to delete an object from the TA secur storage.

This example does not cover all the possibilities of secure storage API
provided to TAs: seeking into a object data stream, manipulating
key material objects instead of raw data objects, etc.

### CODE REVIEW

File: secure_storage/Android.mk [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+LOCAL_CFLAGS += -DANDROID_BUILD
+LOCAL_CFLAGS += -Wall
+
+LOCAL_SRC_FILES += host/main.c
+
+LOCAL_C_INCLUDES := $(LOCAL_PATH)/ta/include \
+		$(OPTEE_CLIENT_EXPORT)/include \
+
+LOCAL_SHARED_LIBRARIES := libteec
+LOCAL_MODULE := tee_example_secure_storage
```
Comment (jbech-linaro) [inline secure_storage/Android.mk]: I think we should prefix all optee_example files with somethings like `ote_<module>` (**O**P-**T**EE **E**xample).

File: secure_storage/Makefile [hunk 2]
```diff
@@ -0,0 +1,15 @@
+export V?=0
```
Comment (jbech-linaro) [inline secure_storage/Makefile]: s/export V?=0/export V ?= 0

File: secure_storage/host/Makefile [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+OBJDUMP = $(CROSS_COMPILE)objdump
+READELF = $(CROSS_COMPILE)readelf
+
+OBJS = main.o
+
+CFLAGS += -Wall -I../ta/include -I$(TEEC_EXPORT)/include -I./include
+#Add/link other required libraries here
+LDADD += -lteec -L$(TEEC_EXPORT)/lib
+
+BINARY = optee_secure_storage
```
Comment (jbech-linaro) [inline secure_storage/host/Makefile]: Same here as for the module in the Android make files, ... `ote_<module>`. At least  we should have the same name in both Android and Linux.

File: secure_storage/host/main.c [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+	printf("- Create and load object in the TA secure storage\n");
+
+	memset(obj1_data, 0xA1, sizeof(obj1_data));
+
+	res = write_secure_object(&ctx, obj1_id,
+				  obj1_data, sizeof(obj1_data));
+	if (res != TEEC_SUCCESS)
+		errx(1, "Failed to create an object in the secure storage");
+
+	printf("- Read back the but in a too small buffer (should fail)\n");
```
Comment (jbech-linaro) [inline secure_storage/host/main.c]: I don't think it is a good idea to show this. This smells a bit more of a test case than just showing how to do things. I would just skip this and go directly with the well sized buffer.

File: secure_storage/host/main.c [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+	if (memcmp(obj1_data, read_data, sizeof(obj1_data)))
+		errx(1, "Unexpected content found in secure storage");
+
+	printf("- Delete the object\n");
+
+	res = delete_secure_object(&ctx, obj1_id);
+	if (res != TEEC_SUCCESS)
+		errx(1, "Failed to delete the object: 0x%x", res);
+
+	printf("- Try to access the object (should fail)\n");
```
Comment (jbech-linaro) [inline secure_storage/host/main.c]: Same here, I would skip this "should fail" section.

File: secure_storage/host/main.c (hunk 4 above)
Comment (jbech-linaro) [reply secure_storage/host/main.c]: However, if you insist of keeping this part, then I think you should at least the `params[1].memref.size` which contains the expected size when `TEEC_ERROR_SHORT_BUFFER` is returned (at least that is what you are doing in the TA).

File: secure_storage/ta/user_ta_header_defines.h [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+#define TA_UUID				TA_SECURE_STORAGE_UUID
+
+#define TA_FLAGS			(TA_FLAG_EXEC_DDR | TA_FLAG_SINGLE_INSTANCE)
+#define TA_STACK_SIZE			(2 * 1024)
+#define TA_DATA_SIZE			(32 * 1024)
+
+#define TA_CURRENT_TA_EXT_PROPERTIES \
+    { "gp.ta.description", USER_TA_PROP_TYPE_STRING, \
+        "Exmaple of TA writing/reading data from its secure storage" }, \
```
Comment (jbech-linaro) [inline secure_storage/ta/user_ta_header_defines.h]: s/Exmaple/Example/

File: secure_storage/ta/user_ta_header_defines.h [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+/*
+ * The name of this file must not be modified
+ */
+
+#ifndef USER_TA_HEADER_DEFINES_H
+#define USER_TA_HEADER_DEFINES_H
+
+#include <secure_storage_ta.h>
+
+#define TA_UUID				TA_SECURE_STORAGE_UUID
```
Comment (jbech-linaro) [inline secure_storage/ta/user_ta_header_defines.h]: Should we use the same idea here? `OTE_<module>_UUID`?

File: secure_storage/host/main.c (hunk 4 above)
Comment (etienne-lms) [reply secure_storage/host/main.c]: I will remove the test case.

File: secure_storage/host/main.c (hunk 5 above)
Comment (etienne-lms) [reply secure_storage/host/main.c]: ok

**Comment by etienne-lms** [comment]
comments addressed.

File: secure_storage/host/main.c [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+				 read_data, sizeof(read_data) - 100);
+	if (res != TEEC_ERROR_SHORT_BUFFER)
+		errx(1, "Expected 'short buffer' error but got 0x%x\n", res);
+
+	printf("- Read back the object (into a well sized buffer)\n");
+
+	res = read_secure_object(&ctx, obj1_id,
+				 read_data, sizeof(read_data));
+	if (res != TEEC_SUCCESS)
+		errx(1, "Failed to create an object in the secure storage");
```
Comment (sahilnxp) [inline secure_storage/host/main.c]: Here Error message should be "Failed to read an object".

**Comment by etienne-lms** [comment]
Rebased on top of https://github.com/linaro-swg/optee_examples/pull/8 and https://github.com/linaro-swg/optee_examples/pull/3.

**Comment by etienne-lms** [comment]
Rebased on master.

File: secure_storage/host/Makefile [hunk 9]
```diff
@@ -0,0 +1,25 @@
+CC      = $(CROSS_COMPILE)gcc
+LD      = $(CROSS_COMPILE)ld
+AR      = $(CROSS_COMPILE)ar
+NM      = $(CROSS_COMPILE)nm
+OBJCOPY = $(CROSS_COMPILE)objcopy
+OBJDUMP = $(CROSS_COMPILE)objdump
+READELF = $(CROSS_COMPILE)readelf
```
Comment (jforissier) [inline secure_storage/host/Makefile]: Only `CC` is needed it seems

Comment (etienne-lms) [reply secure_storage/host/Makefile]: Kept for consistency with other. Will clean all in a later change if better.

File: .gitignore [hunk 10]
```diff
@@ -17,3 +17,4 @@ hello_world/host/optee_example_hello_world
 random/host/optee_example_random
 aes/host/optee_example_aes
 hotp/host/optee_example_hotp
+secure_storage/host/optee_example_secure_storag
```
Comment (jenswi-linaro) [inline .gitignore]: Is an `e` missing at the end?

**Comment by etienne-lms** [comment]
Rebased, tag applied.

File: .gitignore (hunk 10 above)
Comment (etienne-lms) [reply .gitignore]: Good catch. i don't know how I corrupted that!
Thanks.

**Comment by jenswi-linaro** [comment]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by etienne-lms** [comment]
Squashed, commit description fixed, tags applied.
Thanks

**Pull request merged**